### ISSUE TITLE
_num_elements(grad) difference between Tensor and IndexedSlices behavior

### ISSUE THREAD

**Issue by alexeyr** [tensorflow/tensorflow#33681]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux-4.4.0-87-generic-x86_64-with-Ubuntu-16.04-xenial
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.0
- Python version: 3.7
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with: 1. TF 1.0: `python -c "import
tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"` 2. TF 2.0: `python -c
"import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
When training a model (which I can't make public at the moment), I've got an error: `<built-in function len> returned a result with an error set`. Debugging showed the reason was in
https://github.com/tensorflow/tensorflow/blob/0d76b0f988d3d7b06f16da950b3ed95e4ec59e82/tensorflow/python/eager/backprop.py#L615-L624
Namely, my code ended up with `grad` being an instance of `IndexedSlices`, but `grad.values._shape_tuple()` was `(None, 16)`. 

Changing code to
```
    shape_tuple = grad.values._shape_tuple()  # pylint: disable=protected-access
    if shape_tuple is None or None in shape_tuple:
      return 0
    return functools.reduce(operator.mul, shape_tuple, 1)
```
by analogy with the `Tensor` case appears to work and let my model train, but I am not certain whether that's correct.

**Describe the expected behavior**
Either handle the case where `grad.values._shape_tuple()` is/contains `None`, or explicitly throw `ValueError` if it shouldn't be allowed.

**Code to reproduce the issue**
(Will try to do this later)

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by allenlavoie** [comment]
That looks like it's only used here: https://github.com/tensorflow/tensorflow/blob/0250bfdb8c8f0dce3e11c91f5969c994660d21d9/tensorflow/c/eager/tape.h#L785

So safe to return 0; it's just used to estimate when to aggregate gradients, so a performance issue rather than a correctness issue. (CC @alextp )

Want to send a PR with that change? Happy to review. Ideally we'd know when this case is triggered and have a unit test.

**Comment by alexeyr** [comment]
@allenlavoie PR #33912.

**Comment by taylorchu** [comment]
@alexeyr 
I noticed that this path was not always triggered for me. It only happened when I added more layers (more variables, more gradients as a result)
Does this happen for you too? thanks!

**Comment by alexeyr** [comment]
Yes, it's rarely triggered (enough that I still haven't written a good test).

**Comment by alexeyr** [comment]
@taylorchu Actually, if you have a simple enough example triggering it, that would be great.

**Comment by taylorchu** [comment]
weight-share a huge layer more than 4 times? @alexeyr

**Comment by amahendrakar** [comment]
@alexeyr,
Is this still an issue? Could you please upgrade to TensorFlow v2.2 and check if you are still facing the same issue. Thanks!
